backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/storage/backup/HybridBackupCommand.java

```diff
@@ -0,0 +1,295 @@
+        locks.put(getParameters().getVmBackup().getVmId().toString(),
+                LockMessagesMatchUtil.makeLockingPair(LockingGroup.VM, EngineMessage.ACTION_TYPE_FAILED_VM_IS_DURING_BACKUP));
+        return locks;
+
+    }
+
+    @Override
+    protected LockProperties applyLockProperties(LockProperties lockProperties) {
+        return lockProperties.withScope(LockProperties.Scope.Execution);
+    }
+
+    @Override
+    public Map<String, String> getJobMessageProperties() {
+        if (jobProperties == null) {
+            jobProperties = super.getJobMessageProperties();
+            jobProperties.put("vm", getVmName());
+        }
+
+        return super.getJobMessageProperties();
+    }
+
+    private void auditLog(AuditLogType type) {
+        addCustomValue("VmName", getVm().getName());
+        addCustomValue("SnapshotName", StorageConstants.BACKUP_VM_AUTO_GENERATED_SNAPSHOT_DESCRIPTION);
+        auditLogDirector.log(this, type);
+    }
+
+    @Override
+    protected void restoreCommandState() {
+        Guid backupId = getParameters().getVmBackup().getId();
+        VmBackup vmBackup = vmBackupDao.get(backupId);
+        if (CollectionUtils.isEmpty(vmBackup.getDisks())) {
+            vmBackup.setDisks(getDisks());
+        }
+
+        getParameters().setVmBackup(vmBackup);
+    }
+
+    @Override
+    protected List<DiskImage> getDisks() {
+        return diskImageDao.getAllSnapshotsForVmSnapshot(getParameters().getAutoGeneratedSnapshot());
+    }
+
+    private void removeAutoGeneratedSnapshot(CommandContext commandContext, ActionType actionType, VmBackupParameters parameters) {
+        RemoveSnapshotParameters removeSnapshotParameters = new RemoveSnapshotParameters(getParameters().getAutoGeneratedSnapshot(),
+                getVmId());
+        removeSnapshotParameters.setEndProcedure(ActionParametersBase.EndProcedure.COMMAND_MANAGED);
+        if (actionType != null) {
+            removeSnapshotParameters.setParentCommand(actionType);
+        }
+
+        removeSnapshotParameters.setParentParameters(parameters);
+        removeSnapshotParameters.setCorrelationId(getCorrelationId());
+
+        runInternalAction(ActionType.RemoveSnapshot,
+                removeSnapshotParameters,
+                commandContext);
+    }
+
+    private CreateSnapshotForVmParameters buildCreateSnapshotParameters(Guid bitmapId) {
+        CreateSnapshotForVmParameters parameters = new CreateSnapshotForVmParameters(
+                getParameters().getVmId(),
+                StorageConstants.BACKUP_VM_AUTO_GENERATED_SNAPSHOT_DESCRIPTION,
+                false);
+        parameters.setShouldBeLogged(false);
+        parameters.setParentCommand(getActionType());
+        parameters.setParentParameters(getParameters());
+        parameters.setEntityInfo(getParameters().getEntityInfo());
+        parameters.setEndProcedure(ActionParametersBase.EndProcedure.COMMAND_MANAGED);
+        parameters.setNeedsLocking(true);
```

Comment:
not sure that I understand what's going on here in terms of locking - this command locks the disks and the vm exclusively for the duration of the execute phase but create snapshot would also lock the vm?

Reply:
yes, the scope also needs to be extended to Command so CreateSnapshotForVm can release it when it completes

Reply:
makes sense